### PR TITLE
Add Android style logger

### PR DESCRIPTION
**Pull request by SUPERCILEX**
The idea is get logs that look somewhat like Android:
```
06-14 08:53:26.702 3418-3447/com.supercilex.robotscouter.debug W/DynamiteModule: Local module descriptor class for com.google.firebase.auth not found.
```

### CODE REVIEW

File: src/main/java/org/sert2521/gamename/Log.kt [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+    DEBUG,
+    /** Priority constant for the log method; use Log.i. */
+    INFO,
+    /** Priority constant for the log method; use Log.w. */
+    WARN,
+    /** Priority constant for the log method; use Log.e. */
+    ERROR
+}
+
+private val autoTag: String get() {
```
Comment (SUPERCILEX) [inline src/main/java/org/sert2521/gamename/Log.kt]: This needs to be tested, I think it will work but dunno

File: src/main/java/org/sert2521/gamename/Log.kt [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+    Thread.currentThread().stackTrace
+            .filter { it.fileName != "LogKt" }
+            .forEach { return it.className }
+    return ""
+}
+
+fun logV(tag: String? = null, message: String) = log(LogLevel.VERBOSE, tag, message)
```
Comment (SUPERCILEX) [inline src/main/java/org/sert2521/gamename/Log.kt]: Do we want the capitals or just `logv`, `logd`, etc?

File: src/main/java/org/sert2521/gamename/Log.kt [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+fun logI(tag: String? = null, message: String) = log(LogLevel.INFO, tag, message)
+fun logW(tag: String? = null, message: String) = log(LogLevel.WARN, tag, message)
+fun logE(tag: String? = null, message: String) = log(LogLevel.ERROR, tag, message)
+
+private fun log(level: LogLevel, tag: String?, message: String) {
+    if (!isLoggable(level)) return
+    println("$level/${if (tag == null || tag.isEmpty()) autoTag else tag}: $message")
+}
+
+private fun isLoggable(level: LogLevel) = level >= LOG_LEVEL
```
Comment (SUPERCILEX) [inline src/main/java/org/sert2521/gamename/Log.kt]: Here's some Java magic that's rarely used so I'll explain: when you define an enum, Java assigns an `ordinal` variable which is numbered from 0 to n in the order in which you define the constants. Here's the documentation:
http://docs.oracle.com/javase/7/docs/api/java/lang/Enum.html#ordinal()

File: src/main/java/org/sert2521/gamename/Log.kt [hunk 4]
```diff
@@ -0,0 +1,34 @@
+package org.sert2521.gamename
```
Comment (SUPERCILEX) [inline src/main/java/org/sert2521/gamename/Log.kt:1]: In Kotlin, we don't have to define a class if it doesn't need an instance; everything can just be in the file. When it's turned into JVM bytecode, it's your standard utils class with static helper methods. (It'll be called `LogKt` so that's how we use it in Java e.g. `LogKt.logD(...)`. In Kotlin, we just say `logD(...)`)

**Comment by andrewda** [comment]
I've been wondering if we could also add colors. But that all depends on the abilities of the DriverStation console.

**Comment by andrewda** [comment]
Also I kinda like using `Log.v()` instead of `logv()` because it just makes a little more sense to me. Wouldn't this be a good place to use an `object` or Singleton or whatever it's called?

**Comment by SUPERCILEX** [comment]
@andrewda Yeah, that annoyed too, but nah. The point with Kotlin is that you don't need a class if you don't need a class. 😄

Instead I just google an Android logging library and it turn out JetBrains makes one for Android that does pretty much everything so I'm gonna have to use it for Robot Scouter! 😄 Anyway, here's their logging bit: https://github.com/Kotlin/anko/wiki/Anko-Commons-%E2%80%93-Logging

**Comment by SUPERCILEX** [comment]
Yeah, colors would be cool (that's what Android does), but I don't think the drive station supports it. Or at least, I don't know where to start.

**Comment by andrewda** [comment]
I have something in the logging branch that _might_ work but needs testing.

**Comment by SUPERCILEX** [comment]
@andrewda I saw your other branch and I was going to update how we set the logging level to something similiar to what you had done, but then I had an epiphany. We're way too stuck in the past, I'm sick of static constants that require a computer and a rebuild while rushing at competition for something as simple as logging. The much better way to do it IMO is what I did in 947a9ab. We should have been doing it like that years ago so we can simply say "Hey, I want some logging now" instead of the whole rebuild ceremony and stuff.

As for colors, I hope your thing works!!! 😄 Otherwise, 😢

**Comment by andrewda** [comment]
Yea that sounds great! Though I think we _might_ want to use an int that corresponds to the LogLevel enum just to make everything more intuitive.

**Comment by andrewda** [comment]
With the new SmartDashboard I'm working on, people will be able to make their own components, so making a log level drop down would be super easy.

File: src/main/java/org/sert2521/gamename/Log.kt [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+
+fun verbose(tag: String? = null, message: Any) = log(LogLevel.VERBOSE, tag, message)
+fun debug(tag: String? = null, message: Any) = log(LogLevel.DEBUG, tag, message)
+fun info(tag: String? = null, message: Any) = log(LogLevel.INFO, tag, message)
+fun warn(tag: String? = null, message: Any) = log(LogLevel.WARN, tag, message)
+fun error(tag: String? = null, message: Any) = log(LogLevel.ERROR, tag, message)
+
+private fun log(level: LogLevel, tag: String?, message: Any) {
+    if (!isLoggable(level)) return
+    println("$level/${if (tag == null || tag.isEmpty()) autoTag else tag}: $message")
```
Comment (andrewda) [inline src/main/java/org/sert2521/gamename/Log.kt]: Instead of these two lines, does it make more sense to do:

```kotlin
if (isLoggable(level))
    println("$level/${if (tag == null || tag.isEmpty()) autoTag else tag}: $message")
```

That would shorten things up and make the function less complex.

File: src/main/java/org/sert2521/gamename/Drivetrain.kt [hunk 6]
```diff
@@ -10,7 +10,6 @@ import edu.wpi.first.wpilibj.command.Subsystem
  * This is the Drivetrain subsystem, which is in charge of controlling the speeds of the individual
  * talons and converting joystick input to movement speeds.
  */
-
```
Comment (andrewda) [inline src/main/java/org/sert2521/gamename/Drivetrain.kt]: I'm gunna do this in another commit so we don't change files unnecessarily in this PR.

Comment (SUPERCILEX) [reply src/main/java/org/sert2521/gamename/Drivetrain.kt]: Sure, I had to deal with it while resolving merge conflicts.

**Comment by SUPERCILEX** [comment]
> I think we might want to use an int that corresponds to the LogLevel enum just to make everything more intuitive.

Nah, that's the point of `ordinal`s. Also you don't even know it but you aren't using your `priority` property, you're also accidentally [using the `ordinal` constant](https://github.com/SouthEugeneRoboticsTeam/Robot-Base/blob/logging/src/main/java/org/sert2521/gamename/Log.kt#L30). 😄

**Comment by andrewda** [comment]
Alright then, maybe we can at least make it a full string like "verbose" or "error" instead of "v", "e", "d", etc. when entering it in the SmartDashboard – no need to obfuscate things when we could make them clear 👍 .

File: src/main/java/org/sert2521/gamename/Log.kt (hunk 5 above)
Comment (SUPERCILEX) [reply src/main/java/org/sert2521/gamename/Log.kt]: Done.

**Comment by SUPERCILEX** [comment]
@andrewda Yup, I was being lazy. 😄

File: src/main/java/org/sert2521/gamename/Log.kt [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+    /** Priority constant for the log method; use Log.v. */
+    VERBOSE,
+    /** Priority constant for the log method; use Log.d. */
+    DEBUG,
+    /** Priority constant for the log method; use Log.i. */
+    INFO,
+    /** Priority constant for the log method; use Log.w. */
+    WARN,
+    /** Priority constant for the log method; use Log.e. */
+    ERROR
```
Comment (andrewda) [inline src/main/java/org/sert2521/gamename/Log.kt]: > [...] use Log.v.
> [...] use Log.d.
> etc.

These should be changed to whatever we eventually decide the function names should be.

**Review by andrewda — CHANGES_REQUESTED** [review]

File: src/main/java/org/sert2521/gamename/Log.kt (hunk 7 above)
Comment (SUPERCILEX) [reply src/main/java/org/sert2521/gamename/Log.kt]: Yeah, I just copied those constants from the Android source. I don't think we really need documentation, it's pretty clear what those constants are.

File: src/main/java/org/sert2521/gamename/Log.kt [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+            .filter { it.fileName != "LogKt" }
+            .forEach { return it.className }
+    return ""
+}
+
+fun verbose(tag: String? = null, message: Any) = log(LogLevel.VERBOSE, tag, message)
+fun debug(tag: String? = null, message: Any) = log(LogLevel.DEBUG, tag, message)
+fun info(tag: String? = null, message: Any) = log(LogLevel.INFO, tag, message)
+fun warn(tag: String? = null, message: Any) = log(LogLevel.WARN, tag, message)
+fun error(tag: String? = null, message: Any) = log(LogLevel.ERROR, tag, message)
```
Comment (andrewda) [inline src/main/java/org/sert2521/gamename/Log.kt]: Is the tag required? When I type `verbose("Hello")` in IntelliJ it gives me an error saying I need to specify a message.

Comment (SUPERCILEX) [reply src/main/java/org/sert2521/gamename/Log.kt]: Yeah, I'd like to figure that out. You can say `verbose(message = "Hello")`, but because `tag` is first parameter Kotlin assumes you're trying pass in a parameter for `tag` which is why it complains saying you didn't pass in a parameter for `message`. To fix that we would have to have `tag` be the last parameter, but that's weird. Dunno, I don't think `verbose(message = "Hello")` is that bad.

Comment (andrewda) [reply src/main/java/org/sert2521/gamename/Log.kt]: It would add lots of extra complexity, buy we could allow message to be null and, if it is, assume `message = tag` and `tag = null`... But there's gotta be a better way to do that in Kotlin.

Comment (andrewda) [reply src/main/java/org/sert2521/gamename/Log.kt]: Nevermind, looks like it's not possible 🙁  (https://stackoverflow.com/a/34862984/2396125). Moving the tag to the second parameter wouldn't be _that_ weird, and might clean it up a bit. How do they do it at [Anko](https://github.com/Kotlin/anko/wiki/Anko-Commons-%E2%80%93-Logging#trait-like-style) that you linked above? I don't quite understand this or know if it's what we're looking for:

> [...] The methods require you to pass a tag argument. You can eliminate this with using AnkoLogger trait-like interface.

Comment (SUPERCILEX) [reply src/main/java/org/sert2521/gamename/Log.kt]: Lol, we're both real programmers. I literally had that same exact but I was like nah, that's too hacky.

Comment (SUPERCILEX) [reply src/main/java/org/sert2521/gamename/Log.kt]: Yeah, lemme look into how Anko does it

Comment (andrewda) [reply src/main/java/org/sert2521/gamename/Log.kt]: The tag seems like it would be constant throughout a class. I'm guessing that's how Anko does it. You set it per class using their Interface and it uses that tag every time you log in that class.

Comment (SUPERCILEX) [reply src/main/java/org/sert2521/gamename/Log.kt]: @andrewda Yeah, that's how it's done on Android but that's stupid. Anko is doing something really interesting where if you implement their interface, it basically does what my magic method does: https://github.com/Kotlin/anko/blob/master/anko/library/static/commons/src/Logging.kt#L23-L43.

https://github.com/Kotlin/anko/blob/master/anko/library/static/commons/src/Logging.kt#L258

Comment (SUPERCILEX) [reply src/main/java/org/sert2521/gamename/Log.kt]: If my magic method works properly then I don't see why we need to provide a custom tag. I'm just so used to having to provide the tag on Android that my brain was stuck.

Comment (andrewda) [inline src/main/java/org/sert2521/gamename/Log.kt]: Alright. So in Anko if you implement the interface, you can't/shouldn't/wouldn't need to override the magically-generated tag? If so, do we even need a way to specify a tag?

Comment (andrewda) [reply src/main/java/org/sert2521/gamename/Log.kt]: Alright, great!

**Review by andrewda — DISMISSED** [review]
Alright, looks good! If this works on the robot, should be good to merge!

(reminder to myself: test out colors, too)

File: src/main/java/org/sert2521/gamename/Log.kt (hunk 8 above)
Comment (SUPERCILEX) [reply src/main/java/org/sert2521/gamename/Log.kt]: @andrewda Yeah so Anko is using pretty much the most advanced stuff you'll find in Kotlin: extension and inline functions. Once you get it, it's not too hard so I'll try to explain it.

Here's their method:
```kotlin
fun AnkoLogger.debug(...) {
    log(this, message, ...)
}
```

Let's break this apart. `AnkoLogger.debug` is saying we're an extension function meaning we're extending an existing class. This is how Kotlin can provide methods like [`filter`](https://github.com/SouthEugeneRoboticsTeam/Robot-Base/pull/2/files#diff-bbae703ef686b19f885a7f38757aa186R19) which don't exist in the Java version of `List`. So basically `AnkoLogger.debug` is saying we're now inside this class (except you can only access public members because it's compiled down to a static method with the instance as a parameter in Java).

`log(this, ...)` we can do that because remember we are _inside_ `AnkoLogger`.

Now:
```kotlin
private inline fun log(
        logger: AnkoLogger,
        message: Any?,
        thr: Throwable?,
        level: Int,
        f: (String, String) -> Unit,
        fThrowable: (String, String, Throwable) -> Unit) {
    val tag = logger.loggerTag
    if (Log.isLoggable(tag, level)) {
        if (thr != null) {
            fThrowable(tag, message?.toString() ?: "null", thr)
        } else {
            f(tag, message?.toString() ?: "null")
        }
    }
}
```

Here, the `inline` keyword tells the Kotlin compiler to move the code inside the inline function to where it's used to it doesn't have to create an anonymous inner class which is expensive if you're calling it a lot of times. Then `val tag = logger.loggerTag` is getting the class name from that interface you saw. Remember that we were passing in `this` in the previous method which is the class you are inside e.g. the one implementing the interface.

So basically, by implementing that interface, Anko can move itself into your class to pull out its name. The one thing that annoys me with the way they did it is that you actually have to implement their interface which is 💩.

This is a great video explaining these concepts from Android God Jake Wharton: https://youtu.be/A2LukgT2mKc?t=7m59s
If you're a bad person and you don't want to watch the whole video, here's where the hardcore stuff starts: https://youtu.be/A2LukgT2mKc?t=23m46s

I would **_very strongly_** recommend watching the whole thing from where I started it in the first link. You could watch if from the beginning, but it's just going over syntax which you're probably familiar with because JS.

Holly shit that was long! 😄

**Review by andrewda — APPROVED** [review]
